### PR TITLE
Kdb cli rewrite

### PR DESCRIPTION
**Pull request by hannes99**
Added tests and if a command does not exist yet in C, the C++ implementation will be used. And due to #4444 the new implementation does not rely on specload anymore, support for it (`kdb --elektra-spec`) could be added easily if that would still make sense.

## Basics

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add them to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - add a line in `doc/news/_preparation_next_release.md`
  - reformat the code with `scripts/dev/reformat-all`
  - make all unit tests pass
  - fix all memleaks
  - fix the CI itself (or rebase if already fixed)
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if everything is done** and no further pushes are planned by you.

### CODE REVIEW

**Review by markus2330 — CHANGES_REQUESTED** [review]
Please get two other reviews first. Please write more in the release notes of what exactly you did. Please target your new implementation against tools/ckdb, so that both implementations can live side-by-side and users can select via TOOLS which one(s) they want.

**Comment by hannes99** [comment]
@markus2330 currently the C++ implementation is replaced by the C version of a command, so for commands that are not yet implemented(like mounting) in C their C++ implementation is used. In the C version `meta set` for example is a subcommand of `meta` it's a small difference but it would mean that all tests, examples tutorials would have to be in a version where `meta-get` and one where `meta get` is used. No tests, examples etc. were removed, so whatever worked in the C++ version will also work in C. So I don't think it makes sense to have both version in paralell.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Unfortunately still no reviews. Do more reviews for others and specifically ask others to do a review.

Anyway, the PR is a big progress. I am okay with replacing individual commands but without the possibility to compile `kdb` completely without C++, the whole PR only goes the half way. This does not necessarily need to happen within this PR, though.

The main issue is that the release notes do not properly describe what was done. In particular it should point to tutorials/tests that fix bugs etc. What are your contributions? Which advantage to we have if we merge this PR?

File: src/tools/kdb/basename.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#ifndef ELEKTRA_KDB_BASENAME_H
+#define ELEKTRA_KDB_BASENAME_H
+
+#include <kdb.h>
+
+/**
```
Comment (tmakar) [inline src/tools/kdb/basename.h:14]: This is very minor, but what kind of function documentation do we follow in general? For some I see that we are doing documentation in the `*.c` files and some things are documented in `*.h`.

File: src/tools/kdb/get.c [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+	elektraFree ((void *) name);
+
+	KeySet * searchIn = ksNew (0, KS_END);
+	KDB * handle = kdbOpen (NULL, errorKey);
+
+	if (toLookUp == NULL)
+	{
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERRORF (errorKey, "'%s' is not a valid key name", keyName (toLookUp));
+		keyDel (toLookUp);
```
Comment (tmakar) [inline src/tools/kdb/get.c:66]: What about moving all the free memory code to `cleanup` section and check there if the variable is NULL and if not free the memory?

File: src/tools/kdb/get.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ */
+int execGet (KeySet * options, Key * errorKey);
+
+
+// helper functions for the get command
+const char * getCascadingName (const char * str);
+Key * printTrace (KeySet * ks, Key * key, Key * found, elektraLookupFlags options);
+Key * warnOnMeta (KeySet * ks, Key * key, Key * found, elektraLookupFlags options);
+void printOptions (elektraLookupFlags options);
+void setCallback (Key * key, Key * (*f) (KeySet * ks, Key * key, Key * found, elektraLookupFlags flags));
```
Comment (tmakar) [inline src/tools/kdb/get.h:40]: I would suggest describing what kind of callback should be set here. The signature of `*f` is not really intuitive. And maybe also add some description for what the callback is needed.

File: src/tools/kdb/meta-get.c [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+	const char * name = getKeyNameFromOptions (GET_OPTION (options, "keyname"), errorKey, verbose);
+	if (name == NULL)
+	{
+		RETURN (2)
+	}
+
+	const char * metaName = GET_OPTION (options, "metaname");
+
+	Key * toLookUp = keyNew (name, KEY_END);
```
Comment (tmakar) [inline src/tools/kdb/meta-get.c:48]: I would suggest moving `toLookUp` key initialization right before `Key * key = ksLookup (conf, toLookUp, KDB_O_NONE);`. IIRC, I have read in our coding conventions `CODING.md` that we should initialize code as late as possible and it is good practice in general.

File: src/tools/kdb/regex-wrapper.h
Comment (tmakar) [inline src/tools/kdb/regex-wrapper.h:1]: Maybe some description in this header file :thinking: ?

**Review by tmakar — CHANGES_REQUESTED** [review]
Very good work here! I added some notes. For some comments it makes sense to check other occurences.

**Pull request closed without merging**